application/browser/application_security_policy.cc

```diff
@@ -198,32 +198,42 @@ void ApplicationSecurityPolicyCSP::Enforce() {
         // means that can access to all subdomains for 'domain',
         // otherwise, the host of request url should exactly the same
         // as policy.
-        bool subdomains = ((*it).find("*.") == 0);
-        std::string host = subdomains ? (*it).substr(2) : (*it);
+        bool subdomains = (it.find("*.") == 0);
+        const std::string host = subdomains ? it.substr(2) : it;
         AddWhitelistEntry(GURL("http://" + host), subdomains);
         AddWhitelistEntry(GURL("https://" + host), subdomains);
       }
     }
 #endif
   } else {
-    if (!csp_info || csp_info->GetDirectives().empty()) {
-      LOG(ERROR) << "Failed to obtain CSP directives from the manifest";
-      return;
-    }
-    enabled_ = true;
-    const std::map<std::string, std::vector<std::string> >& policies =
-        csp_info->GetDirectives();
-    std::map<std::string, std::vector<std::string> >::const_iterator it =
-        policies.begin();
-    for (; it != policies.end(); ++it) {
-      const std::vector<std::string>& allowed_list = it->second;
-      for (std::vector<std::string>::const_iterator it = allowed_list.begin();
-           it != allowed_list.end(); ++it) {
-        GURL url(*it);
-        if (url.is_valid())
-          AddWhitelistEntry(url, false);
+    if (csp_info && !csp_info->GetDirectives().empty()) {
+      enabled_ = true;
+      const std::map<std::string, std::vector<std::string> >& policies =
+          csp_info->GetDirectives();
+
+      for (const auto& directive : policies) {
+        for (const auto& it : directive.second) {
+          GURL url(it);
+          if (url.is_valid())
+            AddWhitelistEntry(url, false);
+        }
       }
     }
+
+    // scope
+    std::string scope;
+    GURL allowedUrl(ApplicationData::GetBaseURLFromApplicationId(app_->id()));
+    if (app_->data()->GetManifest()->GetString(keys::kScopeKey, &scope) &&
+        !scope.empty()) {
+      enabled_ = true;
+      url::Replacements<char> replacements;
+      replacements.SetPath(scope.c_str(), url::Component(0, scope.length()));
+      allowedUrl = allowedUrl.ReplaceComponents(replacements);
+    }
+    if (allowedUrl.is_valid())
+      AddWhitelistEntry(allowedUrl, false);
```

Comment:
Add a comment that non whilelisted entries are opened externally by the system (e.g. by the browser for HTTP and HTTPS) ?